Layout dns screen

axolotl0212 opened the pull request:
closes #479 

# Screenshots

<img width="954" alt="Screenshot 2020-10-14 at 12 18 27 AM" src="https://user-images.githubusercontent.com/54486231/95888117-456cd700-0db3-11eb-8b9f-b02a2a98f6b6.png">
<img width="1111" alt="Screenshot 2020-10-14 at 12 23 16 AM" src="https://user-images.githubusercontent.com/54486231/95888300-8107a100-0db3-11eb-8dc3-551a25851b71.png">
<img width="1141" alt="Screenshot 2020-10-14 at 12 18 03 AM" src="https://user-images.githubusercontent.com/54486231/95888121-46056d80-0db3-11eb-822e-32634c01e8af.png">
<img width="804" alt="Screenshot 2020-10-14 at 12 19 51 AM" src="https://user-images.githubusercontent.com/54486231/95888100-40a82300-0db3-11eb-8886-364981e50603.png">
<img width="713" alt="Screenshot 2020-10-14 at 12 19 14 AM" src="https://user-images.githubusercontent.com/54486231/95888107-430a7d00-0db3-11eb-8097-d407b4e0debd.png">
<img width="622" alt="Screenshot 2020-10-14 at 12 18 59 AM" src="https://user-images.githubusercontent.com/54486231/95888111-43a31380-0db3-11eb-9847-57186debf437.png">
<img width="799" alt="Screenshot 2020-10-14 at 12 18 43 AM" src="https://user-images.githubusercontent.com/54486231/95888114-443baa00-0db3-11eb-81c0-c430f1bd6605.png">

kiootic reviewed (commented):
- I'm getting this warning in console too:
   ```
   Cache data may be lost when replacing the rawAppConfig field of a App object.

   To address this problem (which is not a bug in Apollo Client), define a custom merge function for the App.rawAppConfig field, so InMemoryCache can safely merge these objects:

     existing: {"id":"test","http":{"public_origin":"https://test.localhost"}}
     incoming: {"id":"test"}

   For more information about these options, please refer to the documentation:

     * Ensuring entity objects have IDs: https://go.apollo.dev/c/generating-unique-identifiers
     * Defining custom merge functions: https://go.apollo.dev/c/merging-non-normalized-objects
   ```
- The add domain button overlapped with the border of the text input, can we move it a bit to the right like the public origin input?
- The public origin input should be a drop down, with the verified domains (with the `https://` prefix) in the list as items.
- Pressing enter while the domain text input is active should add the domain
- Non-custom domains (`isCustom === false`) should not be deletable
- The dismiss animation after deleting domain looks a bit strange


<img width="1440" alt="Screenshot 2020-10-14 at 1 30 23 PM" src="https://user-images.githubusercontent.com/6682519/95948737-b64bd780-0e23-11eb-8b5b-3164b58f3a19.png">

axolotl0212 commented:
@kiootic updated

> I'm getting this warning in console too

seems like you submitted empty string, I haven't seen this error. And we have changed it to dropdown so this should not occur anymore

> The add domain button overlapped with the border of the text input, can we move it a bit to the right like the public origin input?

> The public origin input should be a drop down, with the verified domains (with the https:// prefix) in the list as items.

> Pressing enter while the domain text input is active should add the domain

> Non-custom domains (isCustom === false) should not be deletable

addressed

> The dismiss animation after deleting domain looks a bit strange

As I am always getting error mentioned above, not sure what is wrong with the animation

kiootic commented on portal/src/graphql/portal/mutations/createDomainMutation.ts:
```diff
@@ -0,0 +1,53 @@
+    client,
+    // TODO: backend return whole list of domains so apollo can
+    // automatically update the domain list
+    refetchQueries: ["DomainsQuery"],
```
You should be able to re-select the whole list of domain in the `app` field of payload.

axolotl0212 replied:
It works

kiootic commented on portal/src/graphql/portal/DNSConfigurationScreen.tsx:
```diff
@@ -0,0 +1,598 @@
+}
+
+const DOMAIN_LIST_STYLES: IDetailsListProps["styles"] = {
+  headerWrapper: { marginTop: "-10px" },
```
This makes the header and details list element overlaps. I think we can just omit the 'Domain List' header and remove this style.

axolotl0212 replied:
removed

kiootic commented on portal/src/graphql/portal/DNSConfigurationScreen.tsx:
```diff
@@ -0,0 +1,598 @@
+    const keys = verifiedDomains.map((domain) => {
+      try {
+        // assume scheme included if success
+        const url = new URL(domain.domain);
```
This should always throw an error, since a domain without scheme is not a valid URL. Instead, just appending the `https://` prefix should be good enough for our purpose.

axolotl0212 replied:
Then we need to add verification to make sure user cannot add domain with scheme @kiootic

kiootic replied:
After the changes in #504, adding an invalid domain should properly produce an API error with reason `InvalidDomain`.

kiootic commented on portal/src/graphql/portal/DNSConfigurationScreen.tsx:
```diff
@@ -0,0 +1,598 @@
+    return newDomain !== "";
+  }, [newDomain]);
+
+  // NOTE: if domain is invalid, generic error without extensions is thrown
```
This should be a bug in backend, I'll fix it to produce a proper API error when invalid domain is entered.

axolotl0212 commented:
@kiootic updated and rebased

axolotl0212 commented:
> After the changes in #504, adding an invalid domain should properly produce an API error with reason InvalidDomain.

Handled new error

axolotl0212 commented:
rebased

kiootic reviewed (commented):
I'm getting another GraphQL warning when I'm removing a pending domain:
```
Cache data may be lost when replacing the domains field of a App object.

To address this problem (which is not a bug in Apollo Client), define a custom merge function for the App.domains field, so InMemoryCache can safely merge these objects:

  existing: [{"__ref":"Domain:cGVuZGluZzoxMWEwNWZmZi01ZjkyLTQwZDMtYTQzMC01OTMzOTkwNWUwMjA"},{"__ref":"Domain:dmVyaWZpZWQ6YjgyYTY3MWUtZDQ1NC00ZjdhLThiZjYtNjBlOWZlYTM4ZDU3"}]
  incoming: [{"__ref":"Domain:dmVyaWZpZWQ6YjgyYTY3MWUtZDQ1NC00ZjdhLThiZjYtNjBlOWZlYTM4ZDU3"}]

For more information about these options, please refer to the documentation:

  * Ensuring entity objects have IDs: https://go.apollo.dev/c/generating-unique-identifiers
  * Defining custom merge functions: https://go.apollo.dev/c/merging-non-normalized-objects
```

I suppose we'll need to configure Apollo client with custom merge whenever we introduce a new list field in GraphQL API.

axolotl0212 commented:
@kiootic pushed 1 more commit to fix the warning, could you try to delete again and see if the warning goes away? (as I cannot delete domain at this moment)

kiootic reviewed (approved):
It works fine. By the way, you may want to check #500 for setting up local multi-tenant mode to test custom domain APIs.

axolotl0212 commented:
> It works fine. By the way, you may want to check #500 for setting up local multi-tenant mode to test custom domain APIs.

I will try to follow the guide to setup multi-tenant mode
